assignment3.md

```diff
@@ -0,0 +1,233 @@
+
+(Here the `.v` following the target `note` indicates the part of speech.)
+
+Your objective is to automatically cluster each paraphrase set such that each cluster contains words pertaining to a single *sense*, or meaning, of the target word. Note that a single word from the paraphrase set might belong to one or more clusters.
+
+For evaluation, we take the set of ground truth senses from [WordNet](http://wordnet.princeton.edu).
+
+### Training data
+
+The training data consists of two files -- a vocab file (the input), and a clusters file (to evaluate your output). The vocab file `train_vocab.txt` is formatted such that each line contains one target, its paraphrase set, and the number of ground truth clusters *k*, separated by a `::` symbol:
+
+```
+target.pos :: k :: paraphrase1 paraphrase2 paraphrase3 ...
+```
+
+You can use *k* as input to your clustering algorithm.
+
+The clusters file `train_clusters.txt` contains the ground truth clusters for each target word's paraphrase set, split over *k* lines:
+
+```
+target.pos :: 1 :: paraphrase2 paraphrase6
+target.pos :: 2 :: paraphrase3 paraphrase4 paraphrase5
+...
+target.pos :: k :: paraphrase1 paraphrase9
+```
+
+### Test data
+
+For testing, you will receive only a vocab file containing the test target words and their paraphrase sets. Your job is to create an output file, `test_clusters.txt`, formatted in the same way as `train_clusters.txt`, containing the clusters produced by your system.
+
+## Evaluation
+
+There are many possible ways to evaluate clustering solutions. For this homework we will rely on the paired F-score, which you can read more about in [this paper](https://www.cs.york.ac.uk/semeval2010_WSI/paper/semevaltask14.pdf).
+
+The general idea behind paired F-score is to treat clustering prediction like a classification problem; given a target word and its paraphrase set, we call a *positive instance* any pair of paraphrases that appear together in a ground-truth cluster. Once we predict a clustering solution for the paraphrase set, we similarly generate the set of word pairs such that both words in the pair appear in the same predicted cluster. We can then evaluate our set of predicted pairs against the ground truth pairs using precision, recall, and F-score.
+
+We have provided an evaluation script that you can use when developing your own system. You can run it as follows:
+
+```
+python evaluate.py <GROUND-TRUTH-FILE> <PREDICTED-CLUSTERS-FILE>
+```
+
+Your Tasks
+======================
+You have 3 tasks. 
+
+### 1. Sparse Representations 
+
+Your first task is to generate clusters for the target words in `test_vocab.txt` based on a feature-based (not dense) vector space representation. In this type of VSM, each dimension of the vector space corresponds to a specific feature, such as a context word (see, for example, the term-context matrix described in [Chapter 15.1.2 of Jurafsky & Martin](https://web.stanford.edu/~jurafsky/slp3/15.pdf)). 
+
+For this task, use files:
```

Comment:
We have provided you with skeleton code to get you started. To complete this task, you will need to make updates to the following files:

* `makecooccurrences.py`: to calculate cooccurrence statistics and write word-context matrix to file (this could take a long time (>1 hour) to make plans accordingly)
 
* `question1.py`: to load word-context matrix and cluster.

Reply:
updated.